File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    q = deque([(begin, 0)])
    visited = set()
    visited.add(begin)
    
    while q:
        cur, dist = q.popleft()
        if cur == target:
            return dist
        
        for w in words:
            if w not in visited and w != cur:
                cnt = sum(1 if i != j else 0 for i, j in zip(w, cur))
                if cnt == 1:
                    visited.add(w)
                    q.append((w, dist+1))
        
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

bool is_enable_to_change(string& a, string& b) {
    int cnt = 0;
    for(int i = 0; i < a.size(); i++) {
        if (a[i] != b[i]) cnt ++;
    }
    return cnt == 1;
}

int solution(string begin, string target, vector<string> words) {
    if (find(words.begin(), words.end(), target) == words.end()) return 0;
    
    vector<bool> visited(words.size(), false);
    queue <pair<string, int> > q;
    q.push({begin, 0});
    
    while(!q.empty()) {
        auto [cur, dist] = q.front();
        q.pop();
        
        if (cur == target) return dist;
        
        for (int i = 0; i < words.size(); i++) {
            if (!visited[i] && is_enable_to_change(cur, words[i])) {
                visited[i] = true;
                q.push({words[i], dist + 1});
            }
        }
    }
    
    return 0;
}